handles.go

```diff
@@ -128,13 +139,35 @@ func (bhs *BodyHandles) NewReader(rdr io.ReadCloser) (int, *BodyHandle) {
 	bh.reader = rdr
 	bh.closer = rdr
 	bh.writer = ioutil.Discard
+
+	bhs.lock.Lock()
+	defer bhs.lock.Unlock()
+
 	bhs.handles = append(bhs.handles, bh)
 	return len(bhs.handles) - 1, bh
 }
 
 func (bhs *BodyHandles) NewWriter(w io.Writer) (int, *BodyHandle) {
 	bh := &BodyHandle{}
 	bh.writer = w
+
+	bhs.lock.Lock()
+	defer bhs.lock.Unlock()
+
 	bhs.handles = append(bhs.handles, bh)
 	return len(bhs.handles) - 1, bh
 }
+
+func (bhs *BodyHandles) Close(id int) error {
```

Comment:
as discussed in slack, we'll switch to using a map so that removing an item doesn't invalidate other ids